### DISCUSSION TITLE
What wrong? Next.js Image dont work on server

### DISCUSSION THREAD

**Discussion by Xazgl** [vercel/next.js#50028] [Help] [answered]
### Summary

Next.js Image does not work on the server, but it works on the localhost.

One could think about permissions, because I take images from a third-party domain, BUT I load images in a regular img, both on the server and on the localhost correctly.

The logic is this: I take from the domain side with pictures, images of machines and their characteristics into my database. And already from the database I reflect them in the Image component. Due to the large weight of the image, I applied the sizes attribute.

I don't understand then what the reason may be. I think you need to configure the reception of images from other servers, somewhere other than next.conf.But then it's strange why it takes the image of a character in a regular img.

I read here https://nextjs.org/docs/pages/api-reference/components/image-legacy#remote-patterns

My deploy conf: Node + Next.js/Prisma

Next.js and his < Image /> its work on localhost but on server i have this: enter image description here

In alt there is a regular url, and in src there is already a prefix next and it should choose an optimized size for sizes, but it does not do this. enter image description `here```

Additional information
```
<Image
                                    src={car.picture[0]}
                                    alt={car.picture[0]}
                                    layout="fill"
                                    sizes="(max-width: 750px) 50vw,
                                            (max-width: 828px) 40vw,
                                            (max-width: 1080px) 33vw,
                                            20vw"
                                    loading="lazy"
                                /> 
```
This is the usual < img />.Its work on server or localhost, but images sizes amazing:

```
<img
                            loading="lazy"
                            decoding='async'
                            src={car.picture[0]}
                            className="cardImg">
                        </img> 
```

Added a third-party domain with img in next.config.js:
```

const nextConfig = {
  reactStrictMode: true,
};

module.exports = {
  ...nextConfig,
  images: {
    domains: ['mydomaineImg.com'],
  },
};
```
https://stackoverflow.com/questions/76181016/what-wrong-next-js-image-dont-work-on-server

### Additional information

```js
https://stackoverflow.com/questions/76181016/what-wrong-next-js-image-dont-work-on-server
```

**Comment by icyJoseph** [comment]
When you say, it works on localhost, do you mean in `next dev` or `next build && next start`? 

I wonder if the 3rd party domain has CORS or some other kind of restriction on the caller, but the weird thing is that next/image first sends the request to the next.js server via `_next/` calls, could you try to post some screenshot with what's going on with those calls? They look like this: `/_next/image?url=`, look for them in the network tab.

Are you also copying the `next.config.js` into your production server?

**Reply by Xazgl** [reply]
On localhost next dev. According to the idea, if a third-party domain had CORS with restrictions, then in this case it would not be loaded into the usual <img> tag? ON the production server next.config.js looks the same. Need a screenshot of networks ?

**Reply by icyJoseph** [reply]
Try with build and start on localhost.

Yeah, I want to see whats going on with the `_next/image?url=` calls, they usually contain some information about why it failed to get the image.

> According to the idea, if a third-party domain had CORS with restrictions, then in this case it would not be loaded into the usual  tag?

Well, a regular image tag goes directly to the 3rd party, while the next/image tag goes to the next server first - perhaps the 3rd party checks on origin of the request? and from the server it'd be missing? who knows, we need to debug, test, and invalidate our hypothesis until we hit the reason. Let's not make assumptions.

**Reply by Xazgl** [reply]
My network 
![image](https://github.com/vercel/next.js/assets/86427773/2098f75a-2286-40e3-91ca-a09e3793fe1a)
![image](https://github.com/vercel/next.js/assets/86427773/a8085b6c-6f08-4129-8b3e-4c7f4f4e3c37)
![image](https://github.com/vercel/next.js/assets/86427773/9f29c35e-55a1-43f8-906e-f915855129ba)
![image](https://github.com/vercel/next.js/assets/86427773/3d566405-1f00-4878-b647-a8c8571b5e5b)

My  component in broswer:
`<img alt="https://mydomain/photo/4576/3693699/orig/c74b2ec1d91003efffbf33d62bc52e89.jpg?sid=43" src="/_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=3840&amp;q=75" decoding="async" data-nimg="fill" style="position: absolute; inset: 0px; box-sizing: border-box; padding: 0px; border: none; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;" sizes="(max-width: 750px) 50vw, (max-width: 828px) 40vw, (max-width: 1080px) 33vw, 20vw" srcset="/_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=128&amp;q=75 128w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=256&amp;q=75 256w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=384&amp;q=75 384w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=640&amp;q=75 640w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=750&amp;q=75 750w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=828&amp;q=75 828w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=1080&amp;q=75 1080w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=1200&amp;q=75 1200w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=1920&amp;q=75 1920w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=2048&amp;q=75 2048w, /_next/image?url=https%3A%2F%2Fs1.mydomain.com%2Fphoto%2F4576%2F3693699%2Forig%2Fc74b2ec1d91003efffbf33d62bc52e89.jpg%3Fsid%3D43&amp;w=3840&amp;q=75 3840w">`

My console : 
![image](https://github.com/vercel/next.js/assets/86427773/67713235-8e1f-45a0-b779-cadb7184fc8c)

**Reply by icyJoseph** [reply]
Your network tab has the `fetch/XHR` filter applied :) make sure `All` is selected

**Reply by icyJoseph** [reply]
Alright, but 400 responses, are bad requests, I wonder if the URL encoding of the image is causing some issue? Does the request, in network tab, contain more information? some kind of message? what headers do you see?

**Reply by icyJoseph** [reply]
Aha `"url" parameter is not allowed` 🤨 let's see...

**Reply by icyJoseph** [reply]
There's a test that documents this message, https://github.com/vercel/next.js/blob/canary/test/integration/image-optimizer/test/util.ts#L479

```js
  it('should fail when domain is not defined in next.config.js', async () => {
    const url = `http://vercel.com/button`
    const query = { url, w: ctx.w, q: 100 }
    const opts = { headers: { accept: 'image/webp' } }
    const res = await fetchViaHTTP(ctx.appPort, '/_next/image', query, opts)
    expect(res.status).toBe(400)
    expect(await res.text()).toBe(`"url" parameter is not allowed`)
  })
```

So are you sure you have the `next.config.js` in your production server? How are you deploying the app?

**Reply by Xazgl** [reply]
Yes ,I have this file in the root folder on server now 
![image](https://github.com/vercel/next.js/assets/86427773/2a4661a4-f908-4d88-993f-89ae474f9c09)
Initially, I thought that this permission was missing, I prescribed it, but nothing has changed

**Reply by icyJoseph** [reply]
Could it be that domains is too restrictive, try with:

```js
module.exports = {
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: '**.example.com',
      },
    ],
  },
};
```

It is not a permissions thing anymore, it is the image optimizer that's calculating that the `src` in your next/image is not allowed, as per domains :) try with remote patterns instead.

**Reply by Xazgl** [reply]
It hasn't helped yet)
![image](https://github.com/vercel/next.js/assets/86427773/2d41f9b3-2c2e-4e4c-86bf-67a1840aa290)
I found this code, I tried it yesterday didn't help. I took it here 
https://nextjs.org/docs/pages/api-reference/components/image-legacy#layout

**Reply by icyJoseph** [reply]
Well clearly the issue is with the `src` you are using in production, and how it is not matched in either domains, nor remote patterns. 

Without being able to see the URLs and the configuration, I can't help so much.

This won't change things, but you should just do:

```js
const nextConfig = {
 /* all of the config here */
}

module.exports = nextConfig
```

For what is worth, this is where things are failing https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/match-remote-pattern.ts, at the function `hasMatch`:

We can see that it expects an exact `domain` match, so sub-domain matching doesn't work and that is the reason why remote patterns was introduced, which as you can see is more complex.

By the way, what next.js version do you have? remote patterns was introduced later than domains...

**Reply by Xazgl** [reply]
My next   12.1.0. 
Domain  example : s1.maxposter.ru 
Example image url: https://s1.maxposter.ru/photo/4571/3574366/orig/a7957bd85de1b57031c841b4fc75cccd.jpg?sid=43

My browser full component :  
`<img alt="https://s1.maxposter.ru/photo/4571/3574366/orig/a7957bd85de1b57031c841b4fc75cccd.jpg?sid=43" src="/_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=3840&amp;q=75" decoding="async" data-nimg="fill" style="position: absolute; inset: 0px; box-sizing: border-box; padding: 0px; border: none; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;" sizes="(max-width: 750px) 50vw, (max-width: 828px) 40vw, (max-width: 1080px) 33vw, 20vw" srcset="/_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=128&amp;q=75 128w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=256&amp;q=75 256w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=384&amp;q=75 384w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=640&amp;q=75 640w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=750&amp;q=75 750w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=828&amp;q=75 828w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=1080&amp;q=75 1080w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=1200&amp;q=75 1200w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=1920&amp;q=75 1920w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=2048&amp;q=75 2048w, /_next/image?url=https%3A%2F%2Fs1.maxposter.ru%2Fphoto%2F4571%2F3574366%2Forig%2Fa7957bd85de1b57031c841b4fc75cccd.jpg%3Fsid%3D43&amp;w=3840&amp;q=75 3840w">`



next config 
```

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  ignoreBuildErrors: true,
  ignoreDuringBuilds: true,
}

module.exports = {
  ...nextConfig,
  
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: 's1.maxposter.ru',
      },
    ],
  },
};
```

**Reply by icyJoseph** [reply]
Yeah remote patterns was introduced experimentally in 12.2. https://nextjs.org/blog/next-12-2#remote-patterns-experimental

And became stable on 12.3 -

The domains have to be exact! So your next config must have `s1.example.com` and not just `example.com` - if you need both, include both, or three or four, etc, or upgrade to 12.3 and use remote patterns instead.

You can also delete the above comment if you feel like you want to keep that information private, it is enough now.

**Reply by Xazgl** [reply]
now I 'm going to try to upgrade to 12.3 , I hope my dependencies won 't break , but for 12.1 , I understand there 's nothing like that ?

And you can also help with this topic on next.js https://github.com/vercel/next.js/discussions/50014 ?

**Reply by icyJoseph** [reply]
Yeah in 12.1 there's no remote patterns, only domains, and domains have to be an exact match. You'd need to list all of the possible domains you want to allow.

I think the upgrade should be fine, but let me know if something goes wrong.

**Reply by Xazgl** [reply]
I switched to version 12.3.4, and you need my config now, but not worked 


```
const text Config = {
  react StrictMode: true,
  ignore Build Errors: true,
  ignore During Builds: true,
}

module.exports = {
  ......next Config,
  
  images: {
    remote Patterns: [
      {
        protocol: 'https',
        hostname: 's1.maxposter.ru',
      },
    ],

},
};
```
![image](https://github.com/vercel/next.js/assets/86427773/930fffe6-80ba-4ea6-90ed-21c202220eb8)

**Reply by icyJoseph** [reply]
Could you try with:

```js
const nextConfig = {
  reactStrictMode: true,
  images: {
    remotePatterns: [
      {
        protocol: 'https',
        hostname: '**.maxposter.ru',
      },
    ],
  },
};

module.exports = nextConfig;
```

**Reply by Xazgl** [reply]
Unfortunately, it doesn't help again, I don't have any ideas at all (
![image](https://github.com/vercel/next.js/assets/86427773/fa865b43-fa73-4bb2-bb01-559b6f6d3159)

**Reply by icyJoseph** [reply]
Wow, that's really weird then. 

Let's try to discard that the problem is the next.config file not being loaded:

```js
const nextConfig = {
  reactStrictMode: true,
  images: {
    remotePatterns: [
      {
        protocol: "https",
        hostname: "**.maxposter.ru",
      },
    ],
  },
};

module.exports = () => {
  console.log("Loading Config");
  return nextConfig;
};
```

That should log in your production environment, when you start the server, if it doesn't, then we know the config is not being loaded.

**Reply by Xazgl** [reply]
I put this code. I don't see anything, in the server console
![image](https://github.com/vercel/next.js/assets/86427773/6b2dba68-1601-41df-8f51-04a2e013b64f)
![image](https://github.com/vercel/next.js/assets/86427773/f0acbdda-9f8a-4efc-907d-5a1fc6f1466f)

But in localhost I see : 
![image](https://github.com/vercel/next.js/assets/86427773/c96b550c-b0f9-4544-8093-47b524c73579)


The file itself is on the same level with env and Docker inside the root of the project
![image](https://github.com/vercel/next.js/assets/86427773/e0b06e12-85b8-49c9-82e9-3fa362a32f44)

If it outputs on the local host, then it works on the server as well, is the project the same ?

**Reply by icyJoseph** [reply]
It should only log on the server, during development, during next build time and during next start.

Locally you see it right? But not in your production server? 

This is strange. You are not using standalone mode, you are not doing export and you have a recent Next.js version. What else could be wrong. 

Are you copying the `.next` folder (result of running next build) somewhere else? Are you maybe using a custom next.js server? Very confusing stuff.

**Reply by Xazgl** [reply]
Now I'm trying to figure out what else can be looked at for this reason. In a rare case, I attach my docker and package.json. Essentially I'm running it like this

package.json : 
```
{
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "migrate": "deploy",
    "prisma:generate": "prisma generate",
    "lint": "next lint",
    "xml:init:db": "nodemon --watch ./xmlTasks ./xmlTasks/index.js",
    "seed": "prisma db seed"
  },
  "prisma": {
    "seed": "node prisma/seed.js"
  },
```

My docker:
```
FROM node:lts-alpine AS builder

# Create app directory
WORKDIR /app

# A wildcard is used to ensure both package.json AND package-lock.json are copied
COPY package*.json ./
COPY prisma ./prisma/


# Install app dependencies
RUN npm install
RUN npx prisma generate


COPY . .

RUN npm run build

FROM node:lts-alpine

WORKDIR /app

COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/package*.json ./
COPY --from=builder /app/.next ./.next
COPY --from=builder /app/public ./public
COPY --from=builder /app/prisma ./prisma


# Install Redis
RUN apk add --update redis


# # Stop Redis service
# RUN pkill redis-server && redis-cli shutdown

# Copy Redis configuration file
COPY redis.conf /usr/local/etc/redis/redis.conf  


# Set ownership and permissions for Redis configuration file
RUN chown redis:redis /usr/local/etc/redis/redis.conf

# Set Redis configuration
ENV REDIS_HOST=https://test.opel-arkont-volgograd.ru
ENV REDIS_PORT=6379

ENV NODE_ENV=production



EXPOSE 3000

CMD ["npx", "prisma", "migrate", "deploy", "--preview-feature"]
```

docker-compose: 

```
#arkont subdomen
  app-arkont:
    container_name: app-arkont
    image: arkont-image
    build:
      context: arkont
      dockerfile: Dockerfile
    restart: always
    environment:
      NODE_ENV: production
    networks:
      - arkont-net
    env_file: 
      - arkont/.env
    ports:
      - 7000:3000
    depends_on:
      - "db-arkont"
    command: sh -c "sleep 13 && npx prisma migrate deploy && npm start"
    links:
      - redis:redis   
    # Выше добавляем ссылку на сервис Redis
  
  #  Redis
  redis:
    container_name: redis
    image: redis:latest
    restart: always
    environment:
      NODE_ENV: production
      REDIS_HOST: test.opel-arkont-volgograd.ru
      REDIS_PORT: 6379 
    ports:
      - 6379:6379
    # command: sh -c "sudo systemctl stop redis && redis-server"
    command: redis-server
    networks:
      - arkont-net
      
  db-arkont:
    container_name: db-arkont
    env_file:
      - arkont/.env
    image: postgres:latest
    restart: always
    volumes:
      - db-arkont-data:/var/lib/postgresql/data
    networks:
      - arkont-net


  #adminer 
  adminer:
    container_name: adminer
    image: adminer
    restart: always
    ports:
      - 8080:8080
    environment:
      - PMA_ARBITRARY=1
    networks:
      - arkont-net


volumes:
  db-arkont-data: {}

networks:
  arkont-net:
    name: arkont-net

```

**Reply by icyJoseph** [reply, marked as answer]
In your dockerfile, in the step that copies from builder, you don't copy the `next.config.js` file, it must be present at dev, build, and start time, therefore you must copy it along.

**Reply by Xazgl** [reply]
Thank you very much! I'm a fucking idiot, it was so easy and on the surface, and I didn't find it)

**Comment by smartdeveler** [comment]
For external URL you must add URL to config.next.js and  add **`unoptimized`** prop to Image tag:
 ```
 <Image unoptimized {...rest} />
```

**Reply by DabirRahmani** [reply]
optimizing and caching the optimized image as webp is the core feature of nextjs image. adding `unoptimized ` to tag is a crime
